clang/lib/AST/ExprConstant.cpp

```diff
@@ -2922,7 +2922,7 @@ static bool handleFloatFloatBinOp(EvalInfo &Info, const BinaryOperator *E,
   //   If during the evaluation of an expression, the result is not
   //   mathematically defined [...], the behavior is undefined.
   // FIXME: C++ rules require us to not conform to IEEE 754 here.
-  if (LHS.isNaN()) {
+  if (!Info.getLangOpts().CPlusPlus23 && LHS.isNaN()) {
```

Comment:
Not sure. We land here when the constexpr builtin_* is using a `NAN` argument with a Windows definition which is:
`#define INFINITY ((float)(1e+300 * 1e+300))`
`#define NAN      (-(float)(INFINITY * 0.0F))`

and an` EvalMode= EM_ConstantExpression`. 

`float f1 = __builtin_fminf(NAN, NAN);` returns an `EvaluateFloat() = true` (`EvalMode= EM_IgnoreSideEffects`)
`constexpr float f1 = __builtin_fminf(NAN, NAN);` returns an `EvaluateFloat() = false` (` EvalMode= EM_ConstantExpression`).

May be this is not the right place to do that? Should `keepEvaluatingAfterUndefinedBehavior` return true for C++23 when `EvalMode= EM_ConstantExpression`?

Reply:
May be additional testing should be added with Linux definition of `NAN`.

Reply:
This is basically to allow calls of __bultin_func(NAN, NAN) to be legal.

Reply:
I think we're potentially misinterpreting the standard here regarding NANs. CC @jcranmer-intel @hubert-reinterpretcast 

https://eel.is/c++draft/basic.fundamental#13 is the interesting bit (I think), and I believe it's saying that if the floating-point representation you use has a way to represent a mathematical value then that value is "representable in that type". e.g., if your fp format can support positive/negative infinity, then a positive/negative infinity is considered representable. Note 10 says that more explicitly as: "Since negative and positive infinity are representable in ISO/IEC/IEEE 60559 formats, all real numbers lie within the range of representable values of a floating-point type adhering to ISO/IEC/IEEE 60559."

NANs are representable within ISO/IEC 60559 and are a mathematical result of an operation such as zero divided by zero. So I believe NANs are valid in a constant expression.

MSVC, Clang, and GCC all agree that they're not, but EDG believes they are: https://godbolt.org/z/cxPcKGvP3

Reply:
The plot thickens, this may be a case where it's defined in C but C++ doesn't bother defining much of anything about floating-point and thus it is UB: https://github.com/cplusplus/draft/issues/5407

That said, https://eel.is/c++draft/basic.fundamental#12.sentence-8 seems to make this implementation-defined and it would be reasonable for us to define C++ as following the C semantics in situations where C++ is silent on the behavior.

C is more clear on this point. C23 5.2.5.3.3p8: Floating types shall be able to represent signed zeros or an unsigned zero and all normalized floating point numbers. In addition, floating types may be able to contain other kinds of floating-point numbers, such as subnormal floating-point numbers and unnormalized floating-point numbers, and
values that are not floating-point numbers, such as NaNs and (signed and unsigned) infinities.

Reply:
This is a C function the semantics should reasonably be that of C
https://eel.is/c++draft/c.math#cmath.syn-1

fmin is well defined for Nan https://en.cppreference.com/w/c/numeric/math/fmin
frexp is well defined for NaN https://en.cppreference.com/w/cpp/numeric/math/frexp
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf

Reply:
So, the comment in the code above is not quite correct right? Should that entire condition be removed?

Reply:
https://eel.is/c++draft/library.c#3 says that

>A call to a C standard library function is a non-constant library call ([defns.nonconst.libcall]) if it raises a floating-point exception other than FE_INEXACT.)"

which suggests that floating-point exceptions should generally cause things to fall out of constant expressions. C's Annex F suggests that warnings be emitted for constant expressions that cause floating-point exceptions other than FE_INEXACT (see F.8.2p2).

As for wording, [expr.pre]p4 requires that things be "mathematically defined", except IEEE 754 specifically says (in section 3.2)

> The mathematical structure underpinning the arithmetic in this standard is the extended reals, that is, the set
of real numbers together with positive and negative infinity.

The C++ specification actually does define "mathematically defined" somewhere, in a footnote of https://eel.is/c++draft/sf.cmath.general:

> A mathematical function is mathematically defined for a given set of argument values (a) if it is explicitly defined for that set of argument values, or (b) if its limiting value exists and does not depend on the direction of approach.

(which, as pendants will note, only defines it for functions, not the base operations of C++, and sidesteps the question of what the set of argument values actually is).

It does seem pretty clear from all the standards that a result of infinity is clearly part of the representable types, and it seems a reasonable inference that the result being infinity is "mathematically defined" (division by 0 is permissible in the extended real numbers).

The role of NaNs is... clear as mud. You have to go into the definition of <limits> to find out that NaN is a representable value. IEEE 754 gives multiple specification levels for floating-point, where NaN is a part of some of them, and proceeds to ignore the fact that it did so in the rest of the specification; the definition of operations involving NaN is deferred into an entire separate section. Section 7.2 goes so far as to say that

> The invalid operation exception is signaled if and only if there is no usefully definable result. In these cases
the operands are invalid for the operation to be performed.

Is `0.0 / 0.0` "mathematically defined" per IEEE 754? It's definitely *defined*, but I don't think I can entirely endorse saying that it's "mathematically defined." The rule of thumb that FE_INVALID is signaled when the inputs are non-qNaN but the output is NaN, combined with the above text, makes it suspect that it should be considered a "mathematical" definition. But this also doesn't help with `NaN / NaN`, which is `NaN` but doesn't throw a FE_INVALID.

As noted in the CWG issue Aaron linked to, someone needs to write a paper to clarify this, and... that's going to be me isn't it?

Reply:
My understanding of the intended design was that operations producing NaNs from non-NaNs in C++ constexpr cause the expression to not be constant. That said, NaNs are still in play because of `numeric_limits`.

For `frexp` though, I think there being an "unspecified" value stored for the exponent is a problem where the design could use more clarity. I would think that the operation could still be performed, but the "unspecified" value should probably be considered indeterminate.

Reply:
> My understanding of the intended design was that operations producing NaNs from non-NaNs in C++

What is "operation" here? NaN in arithmetics / conversions / other core language etc are one thing, but
when calling a function which is specifically specified to produce NaN in some cases... it's very unclear to me that the program should be invalid.

If that's the intent, is there anywhere where it would be documented?

Reply:
There's an open CWG issue ([CWG2168](https://cplusplus.github.io/CWG/issues/2168.html)) about this.

Reply:
> There's an open CWG issue ([CWG2168](https://cplusplus.github.io/CWG/issues/2168.html)) about this.

That's referring to INFINITY though. This is more specifically about NAN.

Reply:
> Do we want to do the same for C++23? If not what's exactly an "unspecified" value?

It means that `INT_MAX` and `INT_MIN` is fine (based on the wording). The committee discussion had already decided that the constexpr evaluation need not match the target's runtime in terms of implementation-defined behavior with respect to:
- what cases are errors, and
- how non-error cases of NaNs and infinities, etc. are handled.

Instead, C's Annex F was to specify the behaviour; however, that C's Annex F did not fully specify the result (as in the case in question) was not discussed.

Reply:
I do not think there is any particular reason why this change is C++23 specific.
The question remains whether the change should be made at all.

If the primary rationale here is to facilitate testing using the Windows header definitions, I suspect the answer is that the Windows definition of NaN is not a C++ constant expression.

@jcranmer-intel's comment (https://github.com/llvm/llvm-project/pull/88978/files#discussion_r1576817678), supports the interpretation that operations that result in NaN from non-qNaN inputs are not valid in a constant expression.

Other than wanting to use the Windows header definitions for the tests here, is there a reason why this change belongs in this PR?

Reply:
Removed and edit the SemaCXX/constexpr-math.cpp to reflect the change.